---
title: "Astrocyte_psudotime_analysis"
author: "Ibrahim Saliu"
date: "2025-07-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Xu_pu_Ast_FinalObj <- readRDS("C:/Users/ibrofina/Downloads/Putamen_Astrocyte_FinalObj_dim15_res0.25_ClusterRenamed.rds")
DefaultAssay(Xu_pu_Ast_FinalObj) <- "integrated"

DefaultAssay(Astrocyte_Final) <- "SCT"
# set the order for CTRL, AD and PD
Xu_pu_Ast_FinalObj@meta.data$Genotype <- factor(Xu_pu_Ast_FinalObj@meta.data$Genotype,levels = c("CTRL", "AD", "PD"))

Xu_pu_Ast_FinalObj[["CellTypeID"]] <- Idents(object = Xu_pu_Ast_FinalObj)
DimPlot(Xu_pu_Ast_FinalObj, reduction = "umap", label = TRUE)

# subset Ast-0 + Ast-1 cells
temp_Ast1_Obj <- subset(Xu_pu_Ast_FinalObj, idents = c("Ast-0", "Ast-1"))
DimPlot(temp_Ast1_Obj, reduction = "umap", label = TRUE)

table(temp_Ast1_Obj@meta.data$Genotype, temp_Ast1_Obj@active.ident)

# subset cell for each brain region
Astrocytes_puAst1_CTRL <- subset(temp_Ast1_Obj, subset = Genotype == "CTRL")
DimPlot(Astrocytes_puAst1_CTRL, reduction = "umap", label = TRUE)

Astrocytes_puAst1_AD <- subset(temp_Ast1_Obj, subset = Genotype == "AD") 
DimPlot(Astrocytes_puAst1_AD, reduction = "umap", label = TRUE)

Astrocytes_puAst1_PD <- subset(temp_Ast1_Obj, subset = Genotype == "PD") 
DimPlot(Astrocytes_puAst1_PD, reduction = "umap", label = TRUE)


# Pseudotime analysis of CTRL
Obj_temp <- Astrocytes_puAst1_CTRL
DefaultAssay(Obj_temp) <- "integrated"

# run PCA
Obj_temp <- RunPCA(Obj_temp,  npcs = 15)

#set dims parameter
dim=1:15
Obj_temp <- RunUMAP(Obj_temp, reduction = "pca", dims = dim)

# Visualization
Idents(Obj_temp) <- Obj_temp@meta.data$CellTypeID
DimPlot(Obj_temp, reduction = "umap", label = TRUE)

Astrocytes_puAst1_CTRL <- Obj_temp
DimPlot(Astrocytes_puAst1_CTRL, reduction = "umap", label = TRUE)

png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_cluster.png",   
    width = 4*300,        # 5 x 300 pixels
    height = 3*300,
    res = 300,            # 300 pixels per inch
    pointsize = 12)        # smaller font size
  DimPlot(Astrocytes_puAst1_CTRL, reduction = "umap", label = FALSE, cols =c("coral3", "deepskyblue3"))
dev.off()

png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_By_orig.ident.png",    
  width = 5*300,        # 5 x 300 pixels
  height = 3*300,
  res = 300,            # 300 pixels per inch
  pointsize = 8)        # smaller font size
  DimPlot(Astrocytes_puAst1_CTRL, reduction = "umap", group.by = "orig.ident")
dev.off()

# use low dimention embedding and cell label directly
ReducedDimention_puAst1_CTRL <-  Embeddings(Astrocytes_puAst1_CTRL, reduction = "umap")

# cluster label
CellLabel_puAst1_CTRL <- as.character(Astrocytes_puAst1_CTRL@active.ident)

# use slingshot wrapper function to perform both steps in a single call 
sds_puAst1_CTRL <- slingshot(ReducedDimention_puAst1_CTRL, CellLabel_puAst1_CTRL, start.clus = "Ast-0", omega = TRUE, omega_scale = 1.5, extend = 'n', stretch = 0, approx_points = FALSE)

lapply(sds_puAst1_CTRL@metadata$curves$Lineage1, head) # has one lineage

# get color palette (by cell type)
colourCount <- length(table(Astrocytes_puAst1_CTRL@active.ident)) # number of levels
getPalette <- colorRampPalette(c("coral3", "deepskyblue3"))

# plot all lineage lines in the same graph
png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_curve.png",       width = 3*300,        # 5 x 300 pixels
    height = 3*300,
    res = 300,            # 300 pixels per inch
    pointsize = 5)        # smaller font size
    
  plot(ReducedDimention_puAst1_CTRL, col = getPalette(colourCount)[Astrocytes_puAst1_CTRL@active.ident], asp = 1, pch = 16)
  lines(SlingshotDataSet(sds_puAst1_CTRL), lwd = 3, col = 'black')
dev.off()

#Same analysis for Ast-1 AD and PD cells, as well as Ast-2 CTRL, AD, and PD cells

```{r}

# plot individual gene expression across pseudotime, worked
GeneExpreWithPseutotime_puAst1_CTRL <- t(as.matrix(Expression_puAst1_CTRL))
GeneExpreWithPseutotime_puAst1_CTRL <- merge(GeneExpreWithPseutotime_puAst1_CTRL, slingPseudotime(sds_puAst1_CTRL, na = TRUE), by = "row.names")
rownames(GeneExpreWithPseutotime_puAst1_CTRL) <- GeneExpreWithPseutotime_puAst1_CTRL$Row.names
GeneExpreWithPseutotime_puAst1_CTRL$Row.names <- NULL

# combine with metadata
GeneExpreWithPseutotime_puAst1_CTRL <- merge(GeneExpreWithPseutotime_puAst1_CTRL, Metadata_puAst1_CTRL, by = "row.names")
rownames(GeneExpreWithPseutotime_puAst1_CTRL) <- GeneExpreWithPseutotime_puAst1_CTRL$Row.names
GeneExpreWithPseutotime_puAst1_CTRL$Row.names <- NULL

# take a look of the last 7 columns to make sure data is correct
GeneExpreWithPseutotime_puAst1_CTRL[1:3, (ncol(GeneExpreWithPseutotime_puAst1_CTRL) -7):ncol(GeneExpreWithPseutotime_puAst1_CTRL) ]

# get colors for cell types, use the selected color for Astrocytes subpopulations
#Cell_colors <- rev(names(table(getPalette(colourCount)[Astrocytes_puAst1_CTRL@active.ident])))
Cell_colors <- c("coral3", "deepskyblue3")

# to change lineages: need to change 3 places for gene names
png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_DEGs_Exp_GPC5.png", width = 7*300, height = 5*300, res = 300,  pointsize = 6)
ggplot(GeneExpreWithPseutotime_puAst1_CTRL, aes(x = Lineage1 , y = GPC5, colour = CellTypeID )) + geom_quasirandom(groupOnX = FALSE) + theme_classic() +
    xlab("pseudotime") + ylab("GPC5 expression") + ggtitle("Cells ordered by pseudotime")+scale_colour_manual(values = Cell_colors) + geom_point(size = 3) + geom_smooth(data=GeneExpreWithPseutotime_puAst1_CTRL, aes(x=Lineage1 , y=GPC5), method = 'loess', fill="black",   colour="grey", size=1)
dev.off()

```

# draw heatmap using ComplexHeatmap for each lineage
library(ComplexHeatmap)
library(matrixStats) # for rowSds

# convert data to dense matrix
Data_to_plot_Lineage_puAst1_CTRL <- as.matrix(Data_to_plot_Lineage_puAst1_CTRL)
mat <- Data_to_plot_Lineage_puAst1_CTRL
mat_Zscore_puAst1_CTRL <- (mat-rowMeans(mat))/(rowSds(as.matrix(mat)))[row(mat)]
mat_Zscore_puAst1_CTRL <- mat_Zscore_puAst1_CTRL[complete.cases(as.matrix(mat_Zscore_puAst1_CTRL)), ]
mat_Zscore_puAst1_CTRL[1:2, (ncol(mat_Zscore_puAst1_CTRL) -4):ncol(mat_Zscore_puAst1_CTRL)]

range(mat_Zscore_puAst1_CTRL)

# define color, numbers are the range of the values in the matrix to be ploted
library(circlize) # for colorRamp2
col_fun_heatmap = colorRamp2(c( -3, 0, 3), c( "blue", "white", "red"))

png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_PD_DEGs_Heatmap_km2.png",  
  width = 10*300,        # 5 x 300 pixels
  height = 8*300,
  res = 300,            # 300 pixels per inch
  pointsize = 6)        # smaller font size
HeatmapValue_puAst1_CTRL <- Heatmap(mat_Zscore_puAst1_PD, col = col_fun_heatmap, use_raster = T, name = "Enrichment", cluster_columns = FALSE, show_row_names = F, show_column_names = FALSE, top_annotation = ColumnAnnotation_puAst1_PD, row_km = 3,  row_km_repeats = 100)
HeatmapValue_puAst1_CTRL = draw(HeatmapValue_puAst1_CTRL)
dev.off()

# get row order
HeatmapModules_puAst1_CTRL <- row_order(HeatmapValue_puAst1_CTRL)
lapply(HeatmapModules_puAst1_CTRL, head)
names(HeatmapModules_puAst1_CTRL) <- c("Module2", "Module1" )

###########################
# subset cluster 2 of puAst1_CTRL
mat_Zscore_puAst1_CTRL_Module1 <- as.matrix(mat_Zscore_puAst1_CTRL[HeatmapModules_puAst1_CTRL$Module1,])
png("Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_DEGs_Heatmap_km2_Module1.png",   
  width = 10*300,        # 5 x 300 pixels
  height = 6*300,
  res = 300,            # 300 pixels per inch
  pointsize = 6)        # smaller font size
HeatmapValue_puAst1_CTRL_Module1 <- Heatmap(mat_Zscore_puAst1_CTRL_Module1, col = col_fun_heatmap, name = "Enrichment", cluster_columns = FALSE, show_row_names = TRUE, show_column_names = FALSE, top_annotation = ColumnAnnotation_puAst1_CTRL)
HeatmapValue_puAst1_CTRL_Module1 = draw(HeatmapValue_puAst1_CTRL_Module1)
dev.off()

CoExprModule_puAst1_CTRL_Module1_Genes <- rownames(mat_Zscore_puAst1_CTRL_Module1)
write.csv(list_pseudotime_DEG_puAst1_CTRL[[1]][which(rownames(list_pseudotime_DEG_puAst1_CTRL[[1]]) %in% rownames(mat_Zscore_puAst1_CTRL_Module1)),  ], "Pu_Astrocyte_FinalObj_Dim15_Res0.25_Psuedotime_puAst1_CTRL_DEGs_Heatmap_km2_Module1_Genes.csv")

```{r pressure, echo=FALSE}

#plot heatmap of pseudotime DEGs of homologous activated astrocytes subpopulation across all dataset
#homologous subpopulations from all dataset
PseudotimeDEG_Ast_list <- list(PseudotimeDEG_XJ_Ast1_PU_CTRL, PseudotimeDEG_XJ_Ast1_PU_AD, PseudotimeDEG_XJ_Ast1_PU_PD, PseudotimeDEG_XJ_Ast2_PU_CTRL, PseudotimeDEG_XJ_Ast2_PU_AD, PseudotimeDEG_XJ_Ast2_PU_PD, PseudotimeDEG_LS_Ast1_PFC_CTRL, PseudotimeDEG_LS_Ast1_PFC_AD, PseudotimeDEG_LS_Ast2_PFC_CTRL, PseudotimeDEG_LS_Ast2_PFC_AD, PseudotimeDEG_GA_Ast1_EC_CTRL, PseudotimeDEG_GA_Ast1_EC_AD,  PseudotimeDEG_GA_Ast2_EC_CTRL, PseudotimeDEG_GA_Ast2_EC_AD, PseudotimeDEG_FR_Ast1_ACC_CTRL, PseudotimeDEG_FR_Ast1_ACC_DLBD, PseudotimeDEG_FR_Ast1_ACC_PD, PseudotimeDEG_FR_Ast1_ACC_PDD, PseudotimeDEG_FR_Ast2_ACC_CTRL, PseudotimeDEG_FR_Ast2_ACC_DLBD, PseudotimeDEG_FR_Ast2_ACC_PD, PseudotimeDEG_FR_Ast2_ACC_PDD, PseudotimeDEG_ZY_Ast1_OC_CTRL, PseudotimeDEG_ZY_Ast1_OC_NHD, PseudotimeDEG_ZY_Ast2_OC_CTRL, PseudotimeDEG_ZY_Ast2_OC_NHD,  PseudotimeDEG_GM_Ast1_MTG_CTRL, PseudotimeDEG_GM_Ast1_MTG_AD, PseudotimeDEG_GM_Ast2_MTG_CTRL, PseudotimeDEG_GM_Ast2_MTG_AD, PseudotimeDEG_LH_Ast2_Pu_CTRL, PseudotimeDEG_LH_Ast2_Pu_HD, PseudotimeDEG_LH_Ast2_Cd_CTRL, PseudotimeDEG_LH_Ast2_Cd_HD)

PseudotimeDEG_Ast <- Reduce(function(x, y) merge(x, y, all=TRUE), PseudotimeDEG_Ast_list)

rownames(PseudotimeDEG_Ast) <- PseudotimeDEG_Ast$GeneName
PseudotimeDEG_Ast$GeneName <- NULL
dim(PseudotimeDEG_Ast)
# 17578  x  34

# remove genes that are not expressed in atleast 10 homologous subpopulation out of the 34 homologous subpopulation  
DataToPlot_Ast2 <- PseudotimeDEG_Ast2[rowSums(!is.na(PseudotimeDEG_Ast2)) >= 10, ]
dim(DataToPlot_Ast2)

# assign "NotSig" to NA 
DataToPlot_Ast2[is.na(DataToPlot_Ast2)] <- "NotDEG"

# convert to numerical
DataToPlot_Ast2 <- DataToPlot_Ast2 %>% mutate_all(funs(case_when(
  . == "up" ~ 1,
  . == "down" ~ -1,
  . == "NotDEG"  ~ 0
)))

library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c("mediumblue", "gray98", "orangered2"))

genes.to.label <- c("APOE", "APP", "APOC1", "PARK7", "BAG3", "C3", "CD44", "CLU", "CRYAB", "AQP4", "CST3", "B2M", "SPARCL1", "DPP10", "FOS", "F3", "CRYAB", "GAPDH", "GFAP", "GJA1", "GPC5", "GLUL", "HSPB1", "GAPDH", "ITM2C", "MAOB", "MT3", "MT2A", "NRXN1", "RORA", "S100B", "SLC1A2", "SPARC", "TNC", "VCAN", "VIM", "SIPA1L1")

position.to.lable <- which(rownames(DataToPlot_Ast2) %in% genes.to.label)

genes.to.label <- rownames(DataToPlot_Ast2)[c(which(rownames(DataToPlot_Ast2) %in% genes.to.label))]

# define color annotation 
ColumnAnnotation <- HeatmapAnnotation(region = str_sub(sapply(strsplit(colnames(DataToPlot_Ast2), "_"), function(x){x[2]})), 
                                      subpopulation = str_sub(sapply(strsplit(colnames(DataToPlot_Ast2), "_"), function(x) {x[3]})),
                                      condition = str_sub(sapply(strsplit(colnames(DataToPlot_Ast2), "_"), function(x) {x[4]})),
                      col = list(region = c("pu" = "antiquewhite2", "acc" = "cadetblue2", "ec" = "chartreuse1", "pfc" = "chartreuse4", "oc" = "plum", "cd" = "blue4", "mtg" = "red4"),
                                subpopulation = c("Ast0" = "coral3", "Ast1" = "deepskyblue3", "Ast2" = "goldenrod2", "Ast3" = "gray47", "Ast4" = "mediumturquoise", "Ast5" = "mediumorchid3", "Ast6" = "gold4", "Ast7" = "forestgreen"),
                                condition = c("CTRL" = "green", "AD" = "red", "DLBD" = "orange1", "PD" = "blue", "PDD" = "purple2", "NHD" = "sienna4", "HD" = "midnightblue")), annotation_legend_param = list(grid_height = unit(0.9, "cm"), grid_width = unit(7, "mm"), labels_gp = gpar(fontsize = 16), title_gp = gpar(fontsize = 16, fontface = "bold")))

RowAnnotation = rowAnnotation(genes = anno_mark(at = position.to.lable, labels = genes.to.label ,  labels_gp = gpar(fontsize = 12)))

png("Ast_PseudotimeDEG_sharedByAllData_heatmap_Cluster_by_columnkm2_rowkm3_filtered.png", width = 10*500, height = 8*500, res = 300, pointsize = 8)  
Heatmap(as.matrix(DataToPlot_Ast2), name = "Trend", column_title = "pseudotime DEGs shared by all scRNA-seq studies", column_km_repeats = 100, column_km = 2, row_km = 3, cluster_columns  = TRUE, column_names_rot = 80, column_names_gp = gpar(fontsize = 16), col = col_fun, show_row_names = FALSE,  top_annotation = ColumnAnnotation, right_annotation = RowAnnotation, heatmap_legend_param = list(labels_gp = gpar(fontsize = 14), legend_height = unit(5, "cm"), legend_width = unit(7, "cm"), grid_height = unit(5, "cm"), grid_width = unit(5, "mm"), title_gp = gpar(fontsize = 15, fontface = "bold")))
dev.off()

```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
